backends/hip-gen/ceed-hip-gen-operator-build.cpp

```diff
@@ -26,6 +26,35 @@
 #include "../hip-shared/ceed-hip-shared.h"
 #include "../hip/ceed-hip-compile.h"
 
+
+//------------------------------------------------------------------------------
+// Calculate the block size used for launching the operator kernel 
+//------------------------------------------------------------------------------
+CEED_INTERN int BlockGridCalculate(const CeedInt dim, const CeedInt nelem,
+		                   const CeedInt P1d, const CeedInt Q1d,
+				   CeedInt *block_sizes) {
+  
+  const CeedInt thread1d = CeedIntMax(Q1d, P1d);
+  if (dim==1) {
+    CeedInt elemsPerBlock = 64*thread1d > 256? 256/thread1d : 64;
+    elemsPerBlock = elemsPerBlock>0?elemsPerBlock:1;
+    block_sizes[0] = thread1d;
+    block_sizes[1] = 1;
+    block_sizes[2] = elemsPerBlock;
+  } else if (dim==2) {
+    const CeedInt elemsPerBlock = thread1d<4? 16 : 2;
+    block_sizes[0] = thread1d;
+    block_sizes[1] = thread1d;
+    block_sizes[2] = elemsPerBlock;
+  } else if (dim==3) {
+    const CeedInt elemsPerBlock = thread1d<6? 4 : (thread1d<8? 2 : 1);
+    block_sizes[0] = thread1d;
+    block_sizes[1] = thread1d;
+    block_sizes[2] = elemsPerBlock;
+  }
+  return CEED_ERROR_SUCCESS;
```

Comment:
We still have a function with exactly this name (but `static`) in `ceed-cuda-gen-operator.c`, but that version does a search to minimize "waste". I think these should have different names even though there isn't a symbol collision.

Also, `CEED_INTERN` doesn't have to be repeated here so long as the header with that declaration is included first. (This isn't good or bad, but in most cases it's less surprising to have a single point of authority to determine symbol visibility.)

Reply:
`BlockGridCalculate_HIP` would be fine. I think we need to make do with what we've got for now (can't wait for new HIP interfaces). We probably need to query the wavefront size (ROCm now supports Navi, but it's 32 while MI is 64).

Can we have a crude bound so we can ensure the solutions are safe even if not truly optimal?

Reply:
Do you mean a bound in terms of total threads?  The code as-is (which is what was already in hip-gen, I just pulled it out into a function that can be used by both the compilation and application functions for now) should never have more than 256 threads/block, unless thread1d*thread1d is greater than 256, in which case it will have max thread1d*thread1d threads -- and we'd have to go above p1d/q1d = 32 to have too many.

Reply:
Right, so we can add extra shared memory requirement when calling the function to ensure the solution is safe. If it returns fewer than 256 threads, we could call again with a smaller shared memory placeholder. This may either converge to a fixed point or a limit cycle. In the latter case, we need to choose the safe value.

Reply:
Oh, I see, you meant safe in terms of shared memory usage.

Another problem, whether we have a variable-smem version of `hipOccupancyMaxPotentialBlockSize` or not, is that with hiprtc, using the launch bounds seems to be more important for performance than with nvrtc.  We need the compiled function to call `hipOccupancyMaxPotentialBlockSize` -- but the compiled function will need to include launch bounds -- then if we change the blocksize based on `hipOccupancyMaxPotentialBlockSize`, we would actually need to compile again (and build the code string again, to replace the launch bounds), which then might change the outcome of `hipOccupancyMaxPotentialBlockSize` if the compiler makes changes...meaning this iterative process might become costly in terms of overhead.

Reply:
Yeah, that is indeed frustrating and relevant to https://github.com/ROCm-Developer-Tools/HIP/issues/924

Seems it would be useful to have the compiler's optimization pass choose launch bounds and we query instead of us probing via an entire compilation unit.

Reply:
Let's resolve this thread so we can merge and we can follow up with AMD liaison about the underlying issue. Thanks for your careful work on this.